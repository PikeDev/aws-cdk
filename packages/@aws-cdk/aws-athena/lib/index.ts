// AWS::Athena CloudFormation Resources:
export * from './athena.generated';
export * from './work-group';
export * from './props';
export * from './options';
