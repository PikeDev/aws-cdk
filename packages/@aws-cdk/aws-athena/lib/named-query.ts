import { Construct, IResource, Resource } from '@aws-cdk/core';
import { CfnNamedQuery } from "./athena.generated";

/**
 * Properties for a new named query
 */
export interface NamedQueryProps {
  /**
   * The database to which the query belongs.
   */
  readonly database: string;

  /**
   * The query description.
   *
   * @default - None
   */
  readonly description?: string;

  /**
   * The query name.
   *
   * @default - None
   */
  readonly name?: string;

  /**
   * The SQL query statements that comprise the query.
   */
  readonly queryString: string;
  }

/**
 * Amazon Athena named query
 */
export interface INamedQuery extends IResource {

}

/**
 * Create an Athena named query
 *
 * @resource AWS::Athena::NamedQuery
 */
export class NamedQuery extends Resource implements INamedQuery {

  constructor(scope: Construct, id: string, props: NamedQueryProps) {
    super(scope, id);

    new CfnNamedQuery(this, 'Resource', {
      database: props.database,
      description: props.description,
      name: props.name,
      queryString: props.queryString
    });
  }
}