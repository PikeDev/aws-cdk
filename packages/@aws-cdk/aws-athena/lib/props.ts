export enum ResultEncryptionOption {
    CSE_KMS = 'CSE_KMS',
    SSE_KMS = 'SSE_KMS',
    SSE_S3 = 'SSE_S3'
}