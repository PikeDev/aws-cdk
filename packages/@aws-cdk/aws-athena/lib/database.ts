import * as cfn from '@aws-cdk/aws-cloudformation';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct, Stack } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';
import { FileFormat, RowFormat } from './props';

/**
 * Properties for a new Athena database
 */
export interface DatabaseProps {

  /**
   * The name of the database.
   */
  readonly name: string;

  /**
   * The name of the workgroup this database is in
   *
   * @default None
   */
  readonly workGroupName?: string;
}

/**
 * A new Athena database
 */
export class Database extends Construct {

  /**
   * The name of this database
   */
  public readonly name: string;

  constructor(scope: Construct, id: string, props: DatabaseProps) {
    super(scope, id);

    this.name = props.name;

    const createQueryString = `CREATE DATABASE IF NOT EXISTS ${props.name}`;
    const deleteQueryString = `DROP DATABASE IF EXISTS ${props.name} CASCADE`;

    new cfn.CustomResource(this, 'Resource', {
      provider: QueryProvider.getOrCreate(this),
      resourceType: 'Custom::AthenaDatabase',
      properties: {
        CreateQueryString: createQueryString,
        UpdateQueryString: createQueryString,
        DeleteQueryString: deleteQueryString,
        WorkGroup: props.workGroupName
      }
    });
  }
}

/**
 * Properties for a new Athena table
 */
export interface TableProps {

  /**
   * The name of the table.
   */
  readonly name: string;

  /**
   * The name of the database this table belongs to
   *
   * @default None
   */
  readonly databaseName?: string;

  /**
   * The name of the workgroup this table is in
   *
   * @default None
   */
  readonly workGroupName?: string;

  /**
   * The columns of this table
   */
  readonly columns: Array<[string, string]>;

  /**
   * Specifies the row format of the table and its underlying source data if applicable
   */
  readonly rowFormat: RowFormat;

  /**
   * Specifies the file format for table data
   */
  readonly fileFormat: FileFormat;

  /**
   * The location in Amazon S3 where this table's data resides
   */
  readonly location: s3.IBucket;

  /**
   * The key within the selected `location` bucket
   *
   * @default Root of the bucket
   */
  readonly locationKey?: string;

  /**
   * Whether the data in the `location` bucket is encrypted
   *
   * @default False
   */
  readonly encryptedData?: boolean;
}

/**
 * A new Athena table
 */
export class Table extends Construct {

  /**
   * The name of this table
   */
  public readonly name: string;

  constructor(scope: Construct, id: string, props: TableProps) {
    super(scope, id);

    this.name = props.name;

    const columns = props.columns.map(x => `${x[0]} ${x[1]}`).join(',');

    const createQueryString = `
      CREATE EXTERNAL TABLE IF NOT EXISTS ${props.databaseName}.${props.name} (${columns})
      ROW FORMAT ${props.rowFormat.statement}
      STORED AS ${props.fileFormat.statement}
      LOCATION 's3://${props.location.bucketName}/${props.locationKey ? props.locationKey + '/' : ''}'
      TBLPROPERTIES ('has_encrypted_data' = '${props.encryptedData ? 'true' : 'false'}')`;

    const deleteQueryString = `DROP TABLE ${props.databaseName}.${props.name}`;

    new cfn.CustomResource(this, 'Resource', {
      provider: QueryProvider.getOrCreate(this),
      resourceType: 'Custom::AthenaTable',
      properties: {
        CreateQueryString: createQueryString,
        UpdateQueryString: createQueryString,
        DeleteQueryString: deleteQueryString,
        WorkGroup: props.workGroupName
      }
    });
  }
}

class QueryProvider extends Construct {

  /**
   * Returns the singleton provider.
   */
  public static getOrCreate(scope: Construct): cr.Provider {
    const stack = Stack.of(scope);
    const id = 'com.amazonaws.cdk.athena.query-provider';
    const x = stack.node.tryFindChild(id) as QueryProvider || new QueryProvider(stack, id);
    return x.provider;
  }

  private readonly provider: cr.Provider;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const onEvent = new lambda.Function(this, 'QueryLambda', {
      code: new lambda.InlineCode(handlerCode),
      runtime: lambda.Runtime.PYTHON_3_7,
      handler: 'index.on_event',
      initialPolicy: [
        new iam.PolicyStatement({
          resources: [ '*' ],
          actions: [
            'athena:*'
          ]
        }),
        new iam.PolicyStatement({
          resources: [ '*' ],
          actions: [
            's3:*'
          ]
        }),
        new iam.PolicyStatement({
          resources: [ '*' ],
          actions: [
            'glue:*',
          ]
        })
      ]
    });

    this.provider = new cr.Provider(this, 'QueryProvider', {
      onEventHandler: onEvent,
    });
  }
}

const handlerCode: string = `
import logging
import cfnresponse
import botocore.session
import time

logger = logging.getLogger()
logger.setLevel(logging.INFO)

session = botocore.session.get_session()
client = session.create_client('athena')

def on_event(event, context):
    print(event)
    try:
        physical_id = 'AthenaQuery' + event['RequestId']
        request_type = event['RequestType']

        queryString = ''
        if request_type == 'Create':
            queryString = event['ResourceProperties']['CreateQueryString']
        elif request_type == 'Update':
            queryString = event['ResourceProperties']['UpdateQueryString']
        elif request_type == 'Delete':
            queryString = event['ResourceProperties']['DeleteQueryString']

        workGroup = event['ResourceProperties']['WorkGroup']

        query = client.start_query_execution(
            QueryString=queryString,
            WorkGroup=workGroup
        )

        query_id = query['QueryExecutionId']

        sleep_amount = 2
        max_attempts = 20
        num_attempts = 0
        execution_state = 'RUNNING'

        while execution_state == 'RUNNING' or execution_state == 'QUEUED':
          execution = client.get_query_execution(QueryExecutionId=query_id)
          num_attempts += 1

          execution_state = execution['QueryExecution']['Status']['State']
          if execution_state == 'SUCCEEDED':
            print('Query succeeded')
            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)
            break
          elif execution_state == 'CANCELLED' or execution_state == 'FAILED' or num_attempts >= max_attempts:
            print('Query failed or cancelled: ' + execution['QueryExecution']['Status']['StateChangeReason'])
            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)
            break

          time.sleep(sleep_amount)

    except Exception as e:
        logger.exception(e)
        cfnresponse.send(event, context, cfnresponse.FAILED, {}, physical_id)
`;