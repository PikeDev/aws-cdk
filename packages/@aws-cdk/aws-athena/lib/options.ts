import * as s3 from "@aws-cdk/aws-s3";
import { FileFormat, RowFormat  } from "./props";

/**
 * Options for a new Athena database
 */
export interface DatabaseOptions {

    /**
     * The name of the database.
     */
    readonly name: string;

    /**
     * The tables in this database
     *
     * @default - None
     */
    readonly tables?: TableOptions[];
  }

/**
 * Options for a new Athena table
 */
export interface TableOptions {
    /**
     * The name of the table.
     */
    readonly name: string;

    /**
     * The columns of this table
     */
    readonly columns: Array<[string, string]>;

    /**
     * Specifies the row format of the table and its underlying source data if applicable
     */
    readonly rowFormat: RowFormat;

    /**
     * Specifies the file format for table data
     */
    readonly fileFormat: FileFormat;

    /**
     * The location in Amazon S3 where this table's data resides
     */
    readonly location: s3.IBucket;

    /**
     * The key within the selected `location` bucket
     *
     * @default Root of the bucket
     */
    readonly locationKey?: string;

    /**
     * Whether the data in the `location` bucket is encrypted
     *
     * @default False
     */
    readonly encryptedData?: boolean;
  }