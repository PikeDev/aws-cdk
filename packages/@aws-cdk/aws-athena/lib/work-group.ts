import * as s3 from '@aws-cdk/aws-s3';
import * as kms from '@aws-cdk/aws-kms';
import { Construct, IResource, Resource } from '@aws-cdk/core';
import { ResultEncryptionOption } from './props';
import { CfnWorkGroup } from './athena.generated';

export interface WorkGroupProps {
  /**
   * The workgroup name.
   */
  readonly name: string;

  /**
   * The workgroup description.
   */
  readonly description?: string;

  /**
   * The option to delete the workgroup and its contents even if the workgroup contains any named queries.
   */
  readonly recursiveDeleteOption: boolean;

  /**
   * Whether the workgroup is enabled or not
   *
   * @default True
   */
  readonly enabled?: boolean;

  readonly bytesScannedCutoffPerQuery?: number;

  readonly enforceWorkGroupConfiguration?: boolean;

  readonly publishCloudWatchMetricsEnabled?: boolean;

  readonly requesterPaysEnabled?: boolean;

  /**
   * Which encryption type to use for query results
   *
   * @default Result encryption disabled
   */
  readonly resultEncryptionOption?: ResultEncryptionOption

  /**
   * The KMS key for query result encryption. If specified `resultEncryptionOption`
   * will be set.
   *
   * @default - default master key.
   */
  readonly resultEncryptionKmsKey?: kms.IKey;

  readonly resultOutputLocation: s3.IBucket;
}

export interface IWorkGroup extends IResource {

}

export class WorkGroup extends Resource implements IWorkGroup {
  constructor(scope: Construct, id: string, props: WorkGroupProps) {
    super(scope, id);

    new CfnWorkGroup(this, 'WorkGroup', {
      name: props.name,
      description: props.description,
      recursiveDeleteOption: props.recursiveDeleteOption,
      state: (props.enabled ?? true) ? 'ENABLED' : 'DISABLED',
      workGroupConfiguration: {
        bytesScannedCutoffPerQuery: props.bytesScannedCutoffPerQuery,
        enforceWorkGroupConfiguration: props.enforceWorkGroupConfiguration,
        publishCloudWatchMetricsEnabled: props.publishCloudWatchMetricsEnabled,
        requesterPaysEnabled: props.requesterPaysEnabled,
        resultConfiguration: {
          encryptionConfiguration: props.resultEncryptionOption
            ? {
              encryptionOption: props.resultEncryptionOption,
              kmsKey: props.resultEncryptionKmsKey && props.resultEncryptionKmsKey.keyArn
            }
            : undefined,
          outputLocation: `s3://${props.resultOutputLocation.bucketName}/`
        }
      }
    });
  }
}