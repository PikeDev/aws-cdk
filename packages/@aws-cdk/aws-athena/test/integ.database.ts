import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { Database, FileFormat, RowFormat, Table, WorkGroup } from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-athena-database');

const bucket = new s3.Bucket(stack, 'Bucket');

const workGroup = new WorkGroup(stack, 'WorkGroup', {
    name: 'workgroup',
    recursiveDeleteOption: true,
    resultOutputLocation: bucket
});

new Database(stack, 'Database', {
    name: 'hello',
    workGroupName: workGroup.name
});

new Table(stack, 'Table', {
    name: 'world',
    workGroupName: workGroup.name,
    databaseName: 'hello',
    location: bucket,
    locationKey: 'data',
    columns: [
        ['ColumnA', 'string'],
        ['ColumnB', 'int']
    ],
    rowFormat: RowFormat.delimiters(';'),
    fileFormat: FileFormat.fromClassNames(
        'org.apache.hadoop.mapred.TextInputFormat',
        'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'
    )
});

app.synth();