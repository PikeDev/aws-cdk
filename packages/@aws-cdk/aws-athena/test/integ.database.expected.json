{
  "Resources": {
    "Bucket83908E77": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "WorkGroup26D4F855": {
      "Type": "AWS::Athena::WorkGroup",
      "Properties": {
        "Name": "workgroup",
        "RecursiveDeleteOption": true,
        "State": "ENABLED",
        "WorkGroupConfiguration": {
          "ResultConfiguration": {
            "OutputLocation": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "Bucket83908E77"
                  },
                  "/"
                ]
              ]
            }
          }
        }
      }
    },
    "DatabaseDatabaseQueryLambdaServiceRoleC2044186": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "DatabaseDatabaseQueryLambdaServiceRoleDefaultPolicy37D57D6C": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "athena:*",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "s3:*",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": [
                "glue:GetDatabase",
                "glue:CreateDatabase",
                "glue:DeleteDatabase",
                "glue:GetTables",
                "glue:GetTable",
                "glue:DeleteTable"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "DatabaseDatabaseQueryLambdaServiceRoleDefaultPolicy37D57D6C",
        "Roles": [
          {
            "Ref": "DatabaseDatabaseQueryLambdaServiceRoleC2044186"
          }
        ]
      }
    },
    "DatabaseDatabaseQueryLambda72A9025B": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "\nimport logging\nimport cfnresponse\nimport botocore.session\nimport time\n\nlogger = logging.getLogger()\nlogger.setLevel(logging.INFO)\n\nsession = botocore.session.get_session()\nclient = session.create_client('athena')\n\ndef on_event(event, context):\n    print(event)\n    try:\n        physical_id = 'AthenaQuery' + event['RequestId']\n        request_type = event['RequestType']\n\n        queryString = ''\n        if request_type == 'Create':\n            queryString = event['ResourceProperties']['CreateQueryString']\n        elif request_type == 'Update':\n            queryString = event['ResourceProperties']['UpdateQueryString']\n        elif request_type == 'Delete':\n            queryString = event['ResourceProperties']['DeleteQueryString']\n\n        workGroup = event['ResourceProperties']['WorkGroup']\n\n        query = client.start_query_execution(\n            QueryString=queryString,\n            WorkGroup=workGroup\n        )\n\n        query_id = query['QueryExecutionId']\n\n        sleep_amount = 2\n        max_attempts = 20\n        num_attempts = 0\n        execution_state = 'RUNNING'\n\n        while execution_state == 'RUNNING' or execution_state == 'QUEUED':\n          execution = client.get_query_execution(QueryExecutionId=query_id)\n          num_attempts += 1\n\n          execution_state = execution['QueryExecution']['Status']['State']\n          if execution_state == 'SUCCEEDED':\n            print('Query succeeded')\n            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)\n            break\n          elif execution_state == 'CANCELLED' or execution_state == 'FAILED' or num_attempts >= max_attempts:\n            print('Query failed or cancelled: ' + execution['QueryExecution']['Status']['StateChangeReason'])\n            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)\n            break\n\n          time.sleep(sleep_amount)\n\n    except Exception as e:\n        logger.exception(e)\n        cfnresponse.send(event, context, cfnresponse.FAILED, {}, physical_id)\n"
        },
        "Handler": "index.on_event",
        "Role": {
          "Fn::GetAtt": [
            "DatabaseDatabaseQueryLambdaServiceRoleC2044186",
            "Arn"
          ]
        },
        "Runtime": "python3.7"
      },
      "DependsOn": [
        "DatabaseDatabaseQueryLambdaServiceRoleDefaultPolicy37D57D6C",
        "DatabaseDatabaseQueryLambdaServiceRoleC2044186"
      ]
    },
    "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleB3ED70A9": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleDefaultPolicy869B6F48": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "DatabaseDatabaseQueryLambda72A9025B",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleDefaultPolicy869B6F48",
        "Roles": [
          {
            "Ref": "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleB3ED70A9"
          }
        ]
      }
    },
    "DatabaseDatabaseQueryProviderframeworkonEventC9E9AC4C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3Bucket347EC9ED"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3VersionKeyDE3A9B7E"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3VersionKeyDE3A9B7E"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleB3ED70A9",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "DatabaseDatabaseQueryLambda72A9025B",
                "Arn"
              ]
            }
          }
        },
        "Timeout": 900
      },
      "DependsOn": [
        "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleDefaultPolicy869B6F48",
        "DatabaseDatabaseQueryProviderframeworkonEventServiceRoleB3ED70A9"
      ]
    },
    "DatabaseB269D8BB": {
      "Type": "Custom::AthenaDatabase",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "DatabaseDatabaseQueryProviderframeworkonEventC9E9AC4C",
            "Arn"
          ]
        },
        "CreateQueryString": "CREATE DATABASE IF NOT EXISTS hello",
        "UpdateQueryString": "CREATE DATABASE IF NOT EXISTS hello",
        "DeleteQueryString": "DROP DATABASE IF EXISTS hello CASCADE",
        "WorkGroup": "workgroup"
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "TableTableQueryLambdaServiceRoleB41CAADB": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "TableTableQueryLambdaServiceRoleDefaultPolicy2275D8EF": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "athena:*",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "s3:*",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": [
                "glue:GetDatabase",
                "glue:GetTable",
                "glue:CreateTable",
                "glue:DeleteTable",
                "glue:GetPartitions",
                "glue:GetPartition",
                "glue:DeletePartition"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "TableTableQueryLambdaServiceRoleDefaultPolicy2275D8EF",
        "Roles": [
          {
            "Ref": "TableTableQueryLambdaServiceRoleB41CAADB"
          }
        ]
      }
    },
    "TableTableQueryLambda9772E3A5": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "\nimport logging\nimport cfnresponse\nimport botocore.session\nimport time\n\nlogger = logging.getLogger()\nlogger.setLevel(logging.INFO)\n\nsession = botocore.session.get_session()\nclient = session.create_client('athena')\n\ndef on_event(event, context):\n    print(event)\n    try:\n        physical_id = 'AthenaQuery' + event['RequestId']\n        request_type = event['RequestType']\n\n        queryString = ''\n        if request_type == 'Create':\n            queryString = event['ResourceProperties']['CreateQueryString']\n        elif request_type == 'Update':\n            queryString = event['ResourceProperties']['UpdateQueryString']\n        elif request_type == 'Delete':\n            queryString = event['ResourceProperties']['DeleteQueryString']\n\n        workGroup = event['ResourceProperties']['WorkGroup']\n\n        query = client.start_query_execution(\n            QueryString=queryString,\n            WorkGroup=workGroup\n        )\n\n        query_id = query['QueryExecutionId']\n\n        sleep_amount = 2\n        max_attempts = 20\n        num_attempts = 0\n        execution_state = 'RUNNING'\n\n        while execution_state == 'RUNNING' or execution_state == 'QUEUED':\n          execution = client.get_query_execution(QueryExecutionId=query_id)\n          num_attempts += 1\n\n          execution_state = execution['QueryExecution']['Status']['State']\n          if execution_state == 'SUCCEEDED':\n            print('Query succeeded')\n            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)\n            break\n          elif execution_state == 'CANCELLED' or execution_state == 'FAILED' or num_attempts >= max_attempts:\n            print('Query failed or cancelled: ' + execution['QueryExecution']['Status']['StateChangeReason'])\n            cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_id)\n            break\n\n          time.sleep(sleep_amount)\n\n    except Exception as e:\n        logger.exception(e)\n        cfnresponse.send(event, context, cfnresponse.FAILED, {}, physical_id)\n"
        },
        "Handler": "index.on_event",
        "Role": {
          "Fn::GetAtt": [
            "TableTableQueryLambdaServiceRoleB41CAADB",
            "Arn"
          ]
        },
        "Runtime": "python3.7"
      },
      "DependsOn": [
        "TableTableQueryLambdaServiceRoleDefaultPolicy2275D8EF",
        "TableTableQueryLambdaServiceRoleB41CAADB"
      ]
    },
    "TableTableQueryProviderframeworkonEventServiceRole21B20085": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "TableTableQueryProviderframeworkonEventServiceRoleDefaultPolicy932C83B7": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableTableQueryLambda9772E3A5",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "TableTableQueryProviderframeworkonEventServiceRoleDefaultPolicy932C83B7",
        "Roles": [
          {
            "Ref": "TableTableQueryProviderframeworkonEventServiceRole21B20085"
          }
        ]
      }
    },
    "TableTableQueryProviderframeworkonEvent682A454D": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3Bucket347EC9ED"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3VersionKeyDE3A9B7E"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3VersionKeyDE3A9B7E"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "TableTableQueryProviderframeworkonEventServiceRole21B20085",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "TableTableQueryLambda9772E3A5",
                "Arn"
              ]
            }
          }
        },
        "Timeout": 900
      },
      "DependsOn": [
        "TableTableQueryProviderframeworkonEventServiceRoleDefaultPolicy932C83B7",
        "TableTableQueryProviderframeworkonEventServiceRole21B20085"
      ]
    },
    "TableCD117FA1": {
      "Type": "Custom::AthenaTable",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "TableTableQueryProviderframeworkonEvent682A454D",
            "Arn"
          ]
        },
        "CreateQueryString": {
          "Fn::Join": [
            "",
            [
              "\n      CREATE EXTERNAL TABLE IF NOT EXISTS hello.world (ColumnA string,ColumnB int)\n      ROW FORMAT DELIMITED FIELDS TERMINATED BY ';'\n      STORED AS INPUTFORMAT 'org.apache.hadoop.mapred.TextInputFormat' OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'\n      LOCATION 's3://",
              {
                "Ref": "Bucket83908E77"
              },
              "/data/'\n      TBLPROPERTIES ('has_encrypted_data' = 'false')"
            ]
          ]
        },
        "UpdateQueryString": {
          "Fn::Join": [
            "",
            [
              "\n      CREATE EXTERNAL TABLE IF NOT EXISTS hello.world (ColumnA string,ColumnB int)\n      ROW FORMAT DELIMITED FIELDS TERMINATED BY ';'\n      STORED AS INPUTFORMAT 'org.apache.hadoop.mapred.TextInputFormat' OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'\n      LOCATION 's3://",
              {
                "Ref": "Bucket83908E77"
              },
              "/data/'\n      TBLPROPERTIES ('has_encrypted_data' = 'false')"
            ]
          ]
        },
        "DeleteQueryString": "DROP TABLE world",
        "WorkGroup": "workgroup"
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    }
  },
  "Parameters": {
    "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3Bucket347EC9ED": {
      "Type": "String",
      "Description": "S3 bucket for asset \"082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33a\""
    },
    "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aS3VersionKeyDE3A9B7E": {
      "Type": "String",
      "Description": "S3 key for asset version \"082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33a\""
    },
    "AssetParameters082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33aArtifactHash0DA92241": {
      "Type": "String",
      "Description": "Artifact hash for asset \"082715a6f74a051218c287c9750d2fff07912878d011fad47eb21f33f335a33a\""
    }
  }
}